cohorts/utils.py

```diff
@@ -0,0 +1,124 @@
+        Examples
+        --------- 
+        > print([_clean_column_name(col) for col in ['PD-L1','PD L1','PD L1_']])
+
+    """
+    ## start with input
+    newx = x
+    
+    ## replace meaningful punctuation with text equivalents
+    ## surround each with whitespace to enforce consistent use of _
+    newx = re.sub('<=', ' le ', newx)
+    newx = re.sub('=<', ' le ', newx)
+    newx = re.sub('>=', ' ge ', newx)
+    newx = re.sub('=>', ' ge ', newx)
+    newx = re.sub('<', ' lt ', newx)
+    newx = re.sub('>', ' gt ', newx)
+    newx = re.sub('#', ' num ', newx)
+    
+    ## remove contents within ()
+    if not(keep_parens):
+        newx = re.sub('\([^)]*\)', '', newx)
+    
+    ## replace remaining punctuation/whitespace with _
+    punct_pattern = '[\W_]+'
+    punct_replacement = '_'
+    newx = re.sub(punct_pattern, punct_replacement, newx)
+    
+    ## remove trailing _ if it exists (if last char was punctuation)
+    newx = re.sub('_$', '', newx)
+    return newx
+
+def clean_column_names(cols, keep_parens = True):
+    """
+        Utility script for renaming pandas columns to patsy-friendly names.
+        
+        Revised names have:
+            - punctuation and whitespace -> text or _
+            - all text in lower case 
+
+        Takes a list of column names, returns a dictionary mapping names to revised names.
+
+        If there are any concerns with the conversion, this will print a warning & return original column names. 
+        
+        Parameters 
+        --------------
+            cols (list): list of strings containing column names
+            keep_parens (logical): whether to retain text within parentheses (defalt: True)
+
+        Returns
+        --------
+            dict mapping cols -> newcols
+
+        Example 
+        -------- 
+        > d = {'one' : pd.Series([1., 2., 3.], index=['a', 'b', 'c']),
+              'two' : pd.Series([1., 2., 3., 4.], index=['a', 'b', 'c', 'd']),
+              'PD L1 (value)': pd.Series([1., 2., 3., 4.], index=['a', 'b', 'c', 'd']),
+              'PD L1 (>1)': pd.Series([0., 1., 1., 0.], index=['a', 'b', 'c', 'd']),
+              }
+        > d = pd.DataFrame(d)
+        > d = d.rename(columns = clean_column_names(d.columns))
+        
+        ## observe, by comparison
+        > d2 = d.rename(columns = clean_column_names(d.columns, keep_parens = False))
+
+    """
+    ## clean/replace punctuation
+    newcols = [_clean_column_name(col, keep_parens = keep_parens) for col in cols]
+    
+    ## confirm no duplicates in remaining columns
+    if not(len(newcols) == len(cols)):
+        print('Warning: length of newcols not the same as cols')
+        return dict(zip(cols, cols))
+    
+    if not(len(newcols) == len(set(newcols))):
+        print('Warning: cleanup introduces duplicate names. Please resolve & try again.')
+        return dict(zip(cols, cols))
+    
+    ## modify all column names to lower case
+    newcols_lc = [col.lower() for col in newcols]
```

Comment:
A few reasons:

1. So this can be turned off without having to pass parameter to `_clean_column_names`
2. Separating steps would, I thought, be somewhat cleaner.
3. So uniqueness could be checked before & after case-conversion. Ideally I want to have the non-uniqueness message be more helpful, like "after step X these columns were equivalent - please fix & try again". & actually list the duplicate columns.